File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/CurlUpPower.ts
import GainBlockAction from "../Action/Common/GainBlockAction";
import { RemoveSpecificPowerAction } from "../Action/Common/RemoveSpecificPowerAction";
import { DamageType } from "../DataDefine/DamageType";
import { OnDamageAfterMessage } from "../Events/OnDamageAfterMessage";
import FileHelper from "../FileHelper";
import { LocalizationProperty } from "../Gen/DB/Localization";
import { StringHelper } from "../StringHelper";
import { AbstractPower, PowerType } from "./AbstractPower";


export class CurlUpPower extends AbstractPower{
    public static Id = "Curl Up";
    Icon: string = FileHelper.FormatPath("Powers/minion.png");
    Id: string = "Curl Up";
    Name: string = LocalizationProperty.Read("能力-蜷身");
    private Triggered = false;
    get Type(): PowerType {
        return PowerType.Buff;
    }

    GetDescription(...args: any[]): string {
        return StringHelper.FormatColorString(LocalizationProperty.Read("能力-蜷身描述")).format(this.Amount);
    }

    protected OnDamageAfter(msg: OnDamageAfterMessage) {
        super.OnDamageAfter(msg);
        if(!this.Triggered && msg.info.TotalDamage <this.Owner.Health && msg.info.TotalDamage > 0 && msg.info.DamageType == DamageType.NORMAL){
            this.Flash();
            this.Triggered = true;
            this.AddToBot(new GainBlockAction(this.Owner,this.Amount));
            this.AddToBot(new RemoveSpecificPowerAction(this.Owner,this.Owner,this));
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Power/Buff/CurlUpPower.cs
using ModTheSpire.DataDefine;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using TCG.Config;

namespace ModTheSpire.Power.Buff
{
    public class CurlUpPower : AbstractPower
    {
        public override string Id { get; } = nameof(CurlUpPower);
        public override string Name { get; } = LocalizationProperty.Read("能力-蜷身");
        public override string Icon { get; } = "Powers/minion.png";
        public override PowerType Type { get; } = PowerType.Buff;
        private bool Triggered = false;

        public override string GetDescription(object[] args)
        {
            return string.Format(LocalizationProperty.Read("能力-蜷身描述"), this.Amount);
        }

        protected override void OnDamageAfter(OnDamageAfterMessage msg)
        {
            base.OnDamageAfter(msg);
            if(!this.Triggered && msg.DamageInfo.Total <this.Owner.Health && msg.DamageInfo.Total > 0 && msg.DamageInfo.Type == DamageType.NORMAL){
                this.Flash();
                this.Triggered = true;
                this.AddToBot(new GainBlockAction(this.Owner,this.Amount));
                this.AddToBot(new RemoveSpecificPowerAction(this.Owner,this.Owner,this));
            }
        }
    }
}